js/modules/k6/ws/ws.go

```diff
@@ -0,0 +1,362 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package ws
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"net"
+	"net/http"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/dop251/goja"
+	"github.com/gorilla/websocket"
+	"github.com/loadimpact/k6/js/common"
+	"github.com/loadimpact/k6/lib/metrics"
+	"github.com/loadimpact/k6/lib/netext"
+	"github.com/loadimpact/k6/stats"
+)
+
+type WS struct{}
+
+type Socket struct {
+	ctx           context.Context
+	conn          *websocket.Conn
+	eventHandlers map[string][]goja.Callable
+	scheduled     chan goja.Callable
+	done          chan struct{}
+	shutdownOnce  sync.Once
+}
+
+const writeWait = 10 * time.Second
+
+var (
+	newline = []byte{'\n'}
+	space   = []byte{' '}
+)
+
+func (*WS) Connect(ctx context.Context, url string, args ...goja.Value) *http.Response {
+	rt := common.GetRuntime(ctx)
+	state := common.GetState(ctx)
+
+	setupFn, userTags, header, err := parseArgs(rt, args)
+	if err != nil {
+		// Argument parsing errors should be passed to the user
+		common.Throw(rt, err)
+		return nil
+	}
+
+	tags := map[string]string{
+		"url":   url,
+		"group": state.Group.Path,
+	}
+	// Merge with the user-provided tags
+	for k, v := range userTags {
+		tags[k] = v
+	}
+
+	var bytesRead, bytesWritten int64
+	// Pass a custom net.Dial function to websocket.Dialer that will substitute
+	// the underlying net.Conn with our own TraceConn
+	netDial := func(network, address string) (net.Conn, error) {
+		var err error
+		var conn, traceConn net.Conn
+
+		dialer := netext.NewDialer(net.Dialer{})
+		if conn, err = dialer.DialContext(ctx, network, address); err != nil {
+			return nil, err
+		}
+		if traceConn, err = dialer.TraceConnection(conn, &bytesRead, &bytesWritten); err != nil {
+			return nil, err
+		}
+
+		return traceConn, err
+	}
+
+	wsd := websocket.Dialer{
+		NetDial: netDial,
+		Proxy:   http.ProxyFromEnvironment,
+	}
+
+	start := time.Now()
+	conn, response, connErr := wsd.Dial(url, header)
+	connectionEnd := time.Now()
+	connectionDuration := stats.D(connectionEnd.Sub(start))
+
+	socket := Socket{
+		ctx:           ctx,
+		conn:          conn,
+		eventHandlers: make(map[string][]goja.Callable),
+		scheduled:     make(chan goja.Callable),
+		done:          make(chan struct{}),
+	}
+
+	// Run the user-provided set up function
+	setupFn(goja.Undefined(), rt.ToValue(&socket))
+
+	if connErr != nil {
+		// Pass the error to the user script before exiting immediately
+		socket.handleEvent("error", rt.ToValue(connErr))
+		return response
+	}
+	defer conn.Close()
+
+	tags["status"] = strconv.Itoa(response.StatusCode)
+	tags["subprotocol"] = response.Header.Get("Sec-WebSocket-Protocol")
```

Comment:
All tags should have defaults in the tags map right away, makes it immediately obvious which tags are there + means that they're always present, even through different code paths. Especially important for non-string tags, else simple programming errors might end up with two zero values for them ("0" and "").